### PR TITLE
*: check sc.IgnoreZeroInDate when parsing string or number to date/datetime/timestamp

### LINKED ISSUE #4653: datetime zero in date when select or update with record with bad datetime field
This can be inserted

```sql
TiDB-Server [test]> create table t1  (id int, t datetime not null, primary key(id));
Query OK, 0 rows affected (1.01 sec)

TiDB-Server [test]> insert into t1 values(3, '2017-00-00 00:00:00');
Query OK, 1 row affected (0.00 sec)
```

But query will fail:

```sql
TiDB-Server [test]> select * from t1;
ERROR 1105 (HY000): datetime zero in date
```

And the erroneous record can not be updated either

```sql
TiDB-Server [test]> update t1 set t = '2017-01-01 00:00:00' where id=3;
ERROR 1105 (HY000): datetime zero in date
```

IMHO, tidb should validate before insertion or if record with bad datetime field CAN be inserted tidb should not report error when being selected or updated later.

### PR DESCRIPTION
**Pull request by XuHuaiyu**
fix #4653 

1. add an IgnoreZeroInDate attribute in StatementContext
2. check sc.IgnoreZeroInDate when parse string or number to date/datetime/timestamp

PTAL @coocood @winkyao @zz-jason

### CODE REVIEW

**Comment by shenli** [comment]
Should we also update coprocessor?

**Comment by shenli** [comment]
CI is failed.

**Comment by XuHuaiyu** [comment]
\cc @AndreMouche @hicqu 
Coprocessor may also need to update this.

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -513,7 +513,7 @@ func (b *builtinCastIntAsTimeSig) evalTime(row []types.Datum) (res types.Time, i
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
-	res, err = types.ParseTimeFromNum(val, b.tp.Tp, b.tp.Decimal)
+	res, err = types.ParseTimeFromNum(sc, val, b.tp.Tp, b.tp.Decimal)
```
Comment (zz-jason) [inline expression/builtin_cast.go]: we should check `err` firstly before using `res` ?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -816,7 +817,7 @@ func (b *builtinCastDecimalAsTimeSig) evalTime(row []types.Datum) (res types.Tim
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
-	res, err = types.ParseTime(string(val.ToString()), b.tp.Tp, b.tp.Decimal)
+	res, err = types.ParseTime(sc, string(val.ToString()), b.tp.Tp, b.tp.Decimal)
```
Comment (zz-jason) [inline expression/builtin_cast.go]: ditto

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -969,7 +971,7 @@ func (b *builtinCastStringAsTimeSig) evalTime(row []types.Datum) (res types.Time
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
-	res, err = types.ParseTime(val, b.tp.Tp, b.tp.Decimal)
+	res, err = types.ParseTime(sc, val, b.tp.Tp, b.tp.Decimal)
```
Comment (zz-jason) [inline expression/builtin_cast.go]: ditto

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -1265,7 +1267,7 @@ func (b *builtinCastJSONAsTimeSig) evalTime(row []types.Datum) (res types.Time,
 	if err != nil {
 		return res, false, errors.Trace(err)
 	}
-	res, err = types.ParseTime(s, b.tp.Tp, b.tp.Decimal)
+	res, err = types.ParseTime(sc, s, b.tp.Tp, b.tp.Decimal)
```
Comment (zz-jason) [inline expression/builtin_cast.go]: ditto

File: util/types/time_test.go [hunk 5]
```diff
@@ -28,6 +29,12 @@ type testTimeSuite struct {
 }
 
 func (s *testTimeSuite) TestDateTime(c *C) {
+	sc := mock.NewContext().GetSessionVars().StmtCtx
+	origIgnoreZeroInDate := sc.IgnoreZeroInDate
+	sc.IgnoreZeroInDate = true
+	defer func() {
```
Comment (coocood) [inline util/types/time_test.go]: The `sc` is only used in this test case, we don't need to set it back.

File: util/types/time.go [hunk 6]
```diff
@@ -1314,14 +1317,21 @@ func TimeFromDays(num int64) Time {
...
+func checkDateType(t TimeInternal, allowZeroInDate bool) error {
 	year, month, day := t.Year(), t.Month(), t.Day()
 	if year == 0 && month == 0 && day == 0 {
 		return nil
 	}
 
-	if !allowZeroDate && (month == 0 || day == 0) {
-		return ErrZeroDate
+	if !allowZeroInDate && (month == 0 || day == 0) {
+		errMsg := fmt.Sprintf("%d-00-00", year)
```
Comment (coocood) [inline util/types/time.go]: How about 
```
errMsg = fmt.Sprintf("%04d-%02d-%02d", year, month, day)
```

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @coocood

File: expression/builtin_cast.go [hunk 7]
```diff
@@ -513,7 +513,10 @@ func (b *builtinCastIntAsTimeSig) evalTime(row []types.Datum) (res types.Time, i
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
-	res, err = types.ParseTimeFromNum(val, b.tp.Tp, b.tp.Decimal)
+	res, err = types.ParseTimeFromNum(sc, val, b.tp.Tp, b.tp.Decimal)
+	if err != nil {
+		return res, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_cast.go]:
```go
return res, true, errors.Trace(err)
```

line 525:
```go
return res, false, nil
```

**Comment by coocood** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason 
This commit should be merged with https://github.com/pingcap/tidb-test/pull/368 toghther.

**Pull request merged**